app/models/unsubscribe_key.rb

```diff
@@ -7,23 +7,102 @@ class UnsubscribeKey < ActiveRecord::Base
 
   before_create :generate_random_key
 
-  def self.create_key_for(user, type)
-    if Post === type
-      create(user_id: user.id, unsubscribe_key_type: "topic", topic_id: type.topic_id, post_id: type.id).key
-    else
-      create(user_id: user.id, unsubscribe_key_type: type).key
+  ALL_TYPE = 'all'
+  DIGEST_TYPE = 'digest'
+  TOPIC_TYPE = 'topic'
+
+  class << self
+    def create_key_for(user, type, post: nil)
+      new(user_id: user.id, unsubscribe_key_type: type).tap do |k|
+        if type == TOPIC_TYPE
+          k.topic_id = post.topic_id
+          k.post_id = post.id
+        end
+
+        k.save
+      end.key
+    end
```

Comment:
```suggestion
    def create_key_for(user, type, post: nil)
      unsubscribe_key = new(user_id: user.id, unsubscribe_key_type: type)

      if type == TOPIC_TYPE
        unsubscribe_key.topic_id = post.topic_id
        unsubscribe_key.post_id = post.id
      end

      unsubscribe_key.save!
      unsubscribe_key.key
    end
```

Suggestion as I find using `Object.tap` here much harder to read and not really necessary.